src/app/layout/qgslayoutvaliditychecks.cpp

```diff
@@ -0,0 +1,44 @@
+/***************************************************************************
+                              qgslayoutvaliditychecks.cpp
+                              ---------------------------
+    begin                : November 2018
+    copyright            : (C) 2018 Nyall Dawson
+    email                : nyall dot dawson at gmail dot com
+ ***************************************************************************/
+/***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgslayoutvaliditychecks.h"
+#include "qgsvaliditycheckcontext.h"
+#include "qgslayoutitemmap.h"
+#include "qgslayout.h"
+
+QList<QgsValidityCheckResult> QgsLayoutMapCrsValidityCheck::runCheck( const QgsValidityCheckContext *context, QgsFeedback * ) const
+{
+  QList<QgsValidityCheckResult> results;
+  const QgsLayoutValidityCheckContext *layoutContext = dynamic_cast< const QgsLayoutValidityCheckContext * >( context );
+  if ( !layoutContext )
+    return results;
+
+  QList< QgsLayoutItemMap * > mapItems;
+  layoutContext->layout->layoutItems( mapItems );
+  for ( QgsLayoutItemMap *map : qgis::as_const( mapItems ) )
+  {
+    if ( map->crs().authid() == QStringLiteral( "EPSG:3857" ) )
+    {
+      QgsValidityCheckResult res;
+      res.type = QgsValidityCheckResult::Warning;
+      res.title = tr( "Map projection is misleading" );
+      res.detailedDescription = tr( "The projection for the map item %1 is set to <i>Web Mercator (EPSG:3857)</i> which misrepresents areas and shapes. Consider using an appropriate local projection instead." ).arg( map->displayName() );
```

Comment:
:)

I was expecting this to open much more discussion (maybe it would after merge), and I included this controversial check by default just to raise a discussion about how much "safety belts" should QGIS have, vs how much should we just be a "dumb" tool which can be misused in any way you want.

> There are probably endless warnings we could show regarding CRS - e.g. when using CRS outside of its declared area (like wrong UTM zone), when using unprojected lat/lon CRS, ...

Right. And I'm starting to swing towards the "we SHOULD be warning users about these mistakes" view. (Same with other very common mistakes like field calculating areas/distances when in web mercator).

But that's another discussion, and now that someone at least has seen this I'm going to remove this check and move it to a follow-up PR.

Reply:
I was also expecting a big flame war about web mercator fans and haters :)

Agreed that if we can detect that a user is most likely doing something wrong, it would be useful to warn them - there are lots of possible mistakes even advanced users can easily do. The main question is probably how/where to display the warnings without being too annoying. I am sure everyone still remember this one...

![](http://pixel.nymag.com/imgs/daily/science/2016/05/26/26-clippy.w536.h357.2x.jpg)